graphix/sim/density_matrix.py

```diff
@@ -3,57 +3,102 @@
+        - an iterable of :class:`graphix.states.State` objects
+        - an iterable of iterable of scalars (A 2**n x 2**n numerical density matrix)
+        - a `graphix.statevec.DensityMatrix` object
+        - a `graphix.statevec.Statevector` object
+
+        If `nqubit` is not provided, the number of qubit is inferred from `data` and checked for consistency.
+        If only one :class:`graphix.states.State` is provided and nqubit is a valid integer, initialize the statevector
+        in the tensor product state.
+        If both `nqubit` and `data` are provided, consistency of the dimensions is checked.
+        If a `graphix.statevec.Statevec` or `graphix.statevec.DensityMatrix` is passed, returns a copy.
+
+
+        :param data: input data to prepare the state. Can be a classical description or a numerical input, defaults to graphix.states.BasicStates.PLUS
+        :type data: typing.Union[
+        graphix.states.State,
+        "DensityMatrix",
+        Statevec,
+        typing.Iterable[graphix.states.State],
+        typing.Iterable[numbers.Number],
+        typing.Iterable[typing.Iterable[numbers.Number]],
+        ], optional
+        :param nqubit: number of qubits to prepare, defaults to None
+        :type nqubit: int, optional
         """
-        if data is None:
-            assert nqubit >= 0
-            self.Nqubit = nqubit
-            if plus_state:
-                self.rho = np.ones((2**nqubit, 2**nqubit)) / 2**nqubit
-            else:
-                self.rho = np.zeros((2**nqubit, 2**nqubit))
-                self.rho[0, 0] = 1.0
-        else:
-            if isinstance(data, DensityMatrix):
-                data = data.rho
-            elif isinstance(data, (list, tuple)):
-                data = np.asarray(data, dtype=complex)
-            elif isinstance(data, np.ndarray):
-                pass
-            else:
-                raise TypeError("data must be DensityMatrix, list, tuple, or np.ndarray.")
+        assert nqubit is None or isinstance(nqubit, numbers.Integral) and nqubit >= 0
 
-            assert check_square(data)
-            self.Nqubit = len(data).bit_length() - 1
+        def check_size_consistency(mat):
+            if nqubit is not None and mat.shape != (2**nqubit, 2**nqubit):
+                raise ValueError(
+                    f"Inconsistent parameters between nqubit = {nqubit} and the shape of the provided density matrix = {mat.shape}."
+                )
 
-            self.rho = data
-        assert check_hermitian(self.rho)
-        assert check_unit_trace(self.rho)
+        if isinstance(data, DensityMatrix):
+            check_size_consistency(data)
+            # safe: https://numpy.org/doc/stable/reference/generated/numpy.ndarray.copy.html
+            self.rho = data.rho.copy()
+            self.Nqubit = data.Nqubit
+            return
+        if isinstance(data, typing.Iterable):
+            input_list = list(data)
+            if len(input_list) != 0:
```

Comment:
`len()` on a list does not involve computation by itself, because lists in Python already "know" their length. The costly computation is done at the line above, where `data` (an arbitrary iterator) is coerced into a list: this coercion involves iterating over the all the elements of `data`, computing the length and allocating a new list. All this would be unnecessary if it were just to check that `data` is not empty, but here, in the non-empty case, we need the list of elements anyway (in the body of the `if`), therefore we need to do the coercion anyway.